include/itkBoneMorphometryImageFilter.hxx

```diff
@@ -0,0 +1,266 @@
+ *
+ *  Copyright Insight Software Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkBoneMorphometryImageFilter_hxx
+#define itkBoneMorphometryImageFilter_hxx
+
+#include "itkBoneMorphometryImageFilter.h"
+
+#include "itkImageScanlineIterator.h"
+#include "itkProgressReporter.h"
+#include "itkNeighborhoodAlgorithm.h"
+
+namespace itk
+{
+template< typename TInputImage >
+BoneMorphometryImageFilter< TInputImage >
+::BoneMorphometryImageFilter():
+    m_Threshold(0),
+    m_Pp(0),
+    m_Pl(0),
+    m_PlX(0),
+    m_PlY(0),
+    m_PlZ(0),
+    m_NumVoxels(0),
+    m_NumBoneVoxels(0)
+{
+  this->SetNumberOfRequiredInputs( 1 );
+}
+
+
+template< typename TInputImage >
+void
+BoneMorphometryImageFilter< TInputImage >
+::GenerateInputRequestedRegion()
+{
+  Superclass::GenerateInputRequestedRegion();
+  if ( this->GetInput() )
+    {
+    InputImagePointer image =
+      const_cast< typename Superclass::InputImageType * >( this->GetInput() );
+    image->SetRequestedRegionToLargestPossibleRegion();
+    }
+}
+
+template< typename TInputImage >
+void
+BoneMorphometryImageFilter< TInputImage >
+::EnlargeOutputRequestedRegion(DataObject *data)
+{
+  Superclass::EnlargeOutputRequestedRegion(data);
+  data->SetRequestedRegionToLargestPossibleRegion();
+}
+
+template< typename TInputImage >
+void
+BoneMorphometryImageFilter< TInputImage >
+::AllocateOutputs()
+{
+  // Pass the input through as the output
+  InputImagePointer image =
+    const_cast< TInputImage * >( this->GetInput() );
+
+  this->GraftOutput(image);
+
+  // Nothing that needs to be allocated for the remaining outputs
+}
+
+template< typename TInputImage >
+void
+BoneMorphometryImageFilter< TInputImage >
+::BeforeThreadedGenerateData()
+{
+    ThreadIdType numberOfThreads = this->GetNumberOfThreads();
+
+    m_Pp = 0;
+    m_Pl = 0;
+    m_PlX = 0;
+    m_PlY = 0;
+    m_PlZ = 0;
+
+    // Resize the thread temporaries
+    m_NumVoxels.SetSize(numberOfThreads);
+    m_NumBoneVoxels.SetSize(numberOfThreads);
+    m_NumX.SetSize(numberOfThreads);
+    m_NumY.SetSize(numberOfThreads);
+    m_NumZ.SetSize(numberOfThreads);
+    m_NumXO.SetSize(numberOfThreads);
+    m_NumYO.SetSize(numberOfThreads);
+    m_NumZO.SetSize(numberOfThreads);
+
+    // Resize the thread temporaries
+    m_NumVoxels.Fill(0);
```

Comment:
We don't need to explicitly keep track of `m_NumVoxels` -- it can be determined by the number of voxels in the `RequestedRegion`.